### ISSUE TITLE
How can I clear GPU memory in tensorflow 2?

### ISSUE THREAD

**Issue by HristoBuyukliev** [tensorflow/tensorflow#36465]
### System information
- Custom code; nothing exotic though.
- Ubuntu 18.04
- installed from source (with pip)
- tensorflow version v2.1.0-rc2-17-ge5bf8de
- 3.6
- CUDA 10.1
- Tesla V100, 32GB RAM

I created a model, nothing especially fancy in it. When I create the model, when using nvidia-smi, I can see that tensorflow takes up nearly all of the memory. When I try to fit the model with a small batch size, it successfully runs. When I fit with a larger batch size, it runs out of memory. Nothing unexpected so far. 

However, the only way I can then release the GPU memory is to restart my computer. When I run nvidia-smi I can see the memory is still used, but there is no process using a GPU. Also, If I try to run another model, it fails much sooner. 

Nothing in the first five pages of google results works. (and most solutions are for TF1)

Is there any way to release GPU memory in tensorflow 2?

**Comment by amahendrakar** [comment]
@HristoBuyukliev,
Could you please check [this](https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth) Tensorflow documentation and let us know if it helps. Thanks!

**Comment by HristoBuyukliev** [comment]
@amahendrakar Hi, this is not what I am looking for. Not using up all the memory at once sounds like a useful feature, however I am looking to clear the memory tf has already taken. 

I just tried it out, it doesn't help. I am iteratively increasing batch size, trying to find the biggest one I can use. Once the jupyter kernel crashes, the memory stays taken up.

Additionally, even the advertised functionality does not work. I made a model that had two times fewer parameters, tensorflow still took up 31 out of 32 gigabytes.

**Comment by taborda11** [comment]
Hello @HristoBuyukliev, I had a similar problem when I was iterating over model.predict(), if you are iteratively increasing batch size, try after each batch_size training do `tf.keras.backend.clear_session()`.
That seems to be a case of memory leak in each training.

**Comment by ymodak** [comment]
You may try [limiting gpu memory growth](https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth) in this case.
Put following snippet on top of your code;
```python
import tensorflow as tf
gpus = tf.config.experimental.list_physical_devices('GPU')
tf.config.experimental.set_memory_growth(gpus[0], True)
# your code
```

**Comment by EKami** [comment]
Hi @HristoBuyukliev , this is a very old issue that everyone is facing in TF 1.x as well as TF 2.x, it seems to be a design flaw and the TF team doesn't seem to care about fixing it (I have been facing this issue for more than 2 years now).

What worked well for me was just to run my train/eval in a separate process and wait for it to finish. So when the process finishes the system kills it and releases the GPU resources automatically.
You can achieve this by doing something like:

```
import multiprocessing

process_eval = multiprocessing.Process(target=evaluate, args=(...))
process_eval.start()
process_eval.join()
```

**Comment by HristoBuyukliev** [comment]
@ymodak 
As I also said to amahendrakar:
1. This seems like a nice feature, but not relevant to my problem.
2. Tried it anyway, did not work. 

@taborda11 Thank you for your suggestion, unfortunately it did not work.
@EKami Yes, I figured by now there is no solution. Thank you for your suggestion, I will try it out.

@taborda11 @EKami a teammate of mine found a hacky solution that kind of works:

    for i in $(sudo lsof /dev/nvidia2 | grep python | awk '{print $2}' | sort -u); do sudo kill -9 $i; done

This gets all the python processes that are using GPU2 in my case, and kills them. It works, but is very very ugly and I was hoping for a better way.

**Comment by sanjoy** [comment]
> However, the only way I can then release the GPU memory is to restart my computer. 

How do you exit the TF processes?

This looks like an issue with `nvidia-smi` based on your [last comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582790407).  If `lsof /dev/nvidia2` can find the processes using the GPU then `nvidia-smi` should find them as well.

**Comment by HristoBuyukliev** [comment]
@sanjoy I think that nvidia-smi does not list GPU processes when used within Docker (as in my case)

**Comment by sanjoy** [comment]
> @sanjoy I think that nvidia-smi does not list GPU processes when used within Docker (as in my case)

I see, thanks!

How do you exit the TF processes?

**Comment by mminervini** [comment]
I have also been battling with the issue of releasing GPU memory for quite some time...

My use case is a machine in a production environment with a single Python process that has to serve different types of clients and I need to switch models depending on the service to be provided. Thus, purging previous models from memory is mandatory in this case, otherwise resource exhausted errors appear sooner than later.
With TF 1.x and Keras (when it was separate from TF) I managed to make this work with `keras.backend.clear_session()`. At some point, I decided to finally move to TF 2.x with Keras integrated into TF (`tf.keras`) and then clearing GPU memory apparently became an impossible thing to do! I got the impression that something broke in TF memory management when Keras was integrated into TF.

I tried all combinations of `tf.keras.backend.clear_session()`, `tf.compat.v1.reset_default_graph()`, `gc.collect()`, `close()` the session, `tf.compat.v1.disable_eager_execution()`, and other solutions that I found online, but none of these really solved the issue.

As a last resort, I will try the solution proposed by @EKami to spawn a subprocess every time I need to switch models and I will report on how it goes.
In any case, this introduces inter-process communication and complicates things unnecessarily, so I really hope the TF team will improve GPU memory management and offer a function to really clear the session!

**Comment by mminervini** [comment]
Replying to my own comment...

I implemented the solution based on spawning a subprocess to run Tensorflow code and (as expected) it actually works, because all resources (particularly GPU memory) are released once the subprocess is destroyed.

Of course, there are some drawbacks in terms of implementation complexity, since one has to deal with multi-processing related stuff that otherwise would not be needed, such as inter-process communication or logging from multiple processes.
Performance is also significantly affected, since every subprocess will need to import TF and other modules and load models on the GPU. So, this is definitely not suited for time-critical operations.

**Comment by phiwei** [comment]
@EKami @mminervini 
I have been struggling with this issue for an amount of time that is way beyond reasonable at this point as well... PyTorch did have working solutions for this already two years ago, but I am stuck with TF for now... if you can make the switch, I can warmly recommend it. 

Anyhow, could you point me to a good example / tutorial for the subprocess approach if you know of any?

**Comment by EKami** [comment]
@phiwei Yep we came to the exact same conclusion but it's hard to move big projects which rely so much on TF/Keras to Pytorch. For my future projects, I won't do the same mistake tho and I can clearly see from the papers trends that it's where everyone is heading to. Even the argument of "TF is better suited for production" doesn't hold anymore, in fact we are shooting ourselves in the foot with bugs like this one which even after many years, are still not fixed.

The future is JAX/Pytorch, TF is doomed to be a relic of the past at this rate.

As for the subprocess tutorial, I don't have any to share but the small example I gave here: https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582749350

The bad news is: It seems that this solution doesn't work with TF 2.2 on RTX cards (yet another problem). It works well with RTX cards on TF 1.15.x and non-RTX cards on TF 2.2 (like nvidia T4). It seems to be driver related so maybe with the next driver release for RTX the issue will go away... no idea, we'll see, but at this point, I don't expect much.

**Comment by phiwei** [comment]
@EKami Thanks for the warning, I am in fact using TF 2.2 with an RTX card... I worked with PyTorch a lot two years ago and in my opinion, it was already a very mature tool that actually behaves the way you would want Python code to behave. Something I found very neat was that they by default use dicts for batches, loved that for customising models / handing information through models.

**Comment by mminervini** [comment]
@phiwei I tested the "TF-in-subprocess" approach on a GTX with TF v2.2, but I don't have experience with the RTX series.

As for the how-to, it largely depends on your use case (e.g., grid search training, inference with multiple models, etc.) and the specific features that you want to move inside the subprocesses.

In general, I mainly relied on the Python documentation: [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) for spawning and communicating with the subprocess wrapping the TF code, and [logging cookbook](https://docs.python.org/3/howto/logging-cookbook.html) to implement logging from main process and subprocesses without conflicts.

I hope this helps!

**Comment by houyichaochaochao** [comment]
> @EKami @mminervini
> I have been struggling with this issue for an amount of time that is way beyond reasonable at this point as well... PyTorch did have working solutions for this already two years ago, but I am stuck with TF for now... if you can make the switch, I can warmly recommend it.
> 
> Anyhow, could you point me to a good example / tutorial for the subprocess approach if you know of any?
so maybe like this,hope to help you!

`import multiprocessing
p = miltiprocessing.Process(target=your_train_task,args=(a,)
p.start()
p.join()`

**Comment by Santosh-Gupta** [comment]
I am wondering if there has been any progress. I try to increase my batch size to see how much my model can handle, but then when I reach that point, any batch size gives an OOM error.

**Comment by GatGit12** [comment]
I'd be interested in that, too. 
Especially since the provided methods (set_memory_growth and per_process_gpu_memory_fraction) are not sufficient enough to really release GPU Memory completely.

Is there a way in the C-API to release everything completely?

**Comment by leonjovanovic** [comment]
Same here, wondering if they are even working on that..

**Comment by andyrevell** [comment]
Interesting - @EKami and @phiwei are right. I think I am going to get my lab to make the switch to Pytorch. One PhD student made the switch, but I will advocate for the rest of to do the same. These sort of issues, among others, are really problematic for tensorflow, and if it is true that Pytorch is as mature as I've been hearing, then I won't look back

**Comment by phiwei** [comment]
@andyrevell I have made the switch now and have not regretted it one minute, my productivity has increased a lot. I can especially recommend the combination with ray tune: https://docs.ray.io/en/latest/tune/
it is a bit tricky to figure out how to best integrate this but is a great tool for HP optimisation once it is running.

**Comment by robotoil** [comment]
I use this to release memory. I run it at the start and end of the process. It's important to run this at the end of your program even if the program failed. If you start a new session without having released the memory you will not get the memory back when you run this again. The only way I have found to get memory that has already been allocated back in that situation is to reboot.

from numba import cuda
cuda.select_device(0)
cuda.close()
print('CUDA memory released: GPU0')

**Comment by pryanga-wk** [comment]
I usually run into GPU memory issue when I train a model in one notebook  and move to train a new model in another notebook. 

@robotoil I placed your code snippet before training a tensorflow tf2.0) model in jupyter notebook moving from another jupyter notebook. The following code snippet didnt help :(

```
from numba import cuda
cuda.select_device(0)
cuda.close()
print('CUDA memory released: GPU0')
```

**Comment by GF-Huang** [comment]
Any progress?

**Comment by GF-Huang** [comment]
> I use this to release memory. I run it at the start and end of the process. It's important to run this at the end of your program even if the program failed. If you start a new session without having released the memory you will not get the memory back when you run this again. The only way I have found to get memory that has already been allocated back in that situation is to reboot.
> 
> from numba import cuda
> cuda.select_device(0)
> cuda.close()
> print('CUDA memory released: GPU0')

Once run this, you can not `fit` again, it stucks.

**Comment by lukas-a-olson** [comment]
This did the trick for me: [https://stackoverflow.com/questions/15197286/how-can-i-flush-gpu-memory-using-cuda-physical-reset-is-unavailable/46597252#46597252](https://stackoverflow.com/questions/15197286/how-can-i-flush-gpu-memory-using-cuda-physical-reset-is-unavailable/46597252#46597252)

**Comment by smougel** [comment]
`nvidia-smi`
Wed Nov 18 15:12:04 2020      

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 455.28       Driver Version: 455.28       CUDA Version: 11.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  GeForce RTX 206...  Off  | 00000000:01:00.0 Off |                  N/A |
| N/A   53C    P0    31W /  N/A |   2815MiB /  5934MiB |      4%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      2975      G   /usr/lib/xorg/Xorg                295MiB |
|    0   N/A  N/A      3160      G   /usr/bin/gnome-shell               89MiB |
|    0   N/A  N/A     17302      G   /usr/lib/firefox/firefox            2MiB |
|    0   N/A  N/A     17309      G   /usr/lib/firefox/firefox            2MiB |
|    0   N/A  N/A     17327      G   /usr/lib/firefox/firefox           18MiB |
|    0   N/A  N/A     17354      G   /usr/lib/firefox/firefox            2MiB |
|    0   N/A  N/A     17381      G   /usr/lib/firefox/firefox            2MiB |
|    0   N/A  N/A     17399      G   /usr/lib/firefox/firefox            2MiB |
|    0   N/A  N/A     47887      G   ...gAAAAAAAAA --shared-files       27MiB |
|    0   N/A  N/A     97563      G   ...AAAAAAAA== --shared-files       25MiB |
|    0   N/A  N/A    125022      C   python                           2339MiB |
+-----------------------------------------------------------------------------+
```
and next

`kill -9 125022`

**Comment by louis925** [comment]
> 
> `kill -9 125022`
No, I think the main point of this issue is that we want a way to clear the tensorflow gpu memory usage while keeping the main python process.

**Comment by sanjoy** [comment]
CC @imintz

**Comment by uzi0espil** [comment]
The inability to clear the GPU RAM after clearing the session in production is a huge problem. Unfortunately, I am unable to apply the Process trick because I am running the model in Celery worker which is a daemonic process and it is not allowed to have children. and freeing the RAM using numba.cuda will put the GPU in an unhealthy state.  There should be a way to cleanly clear the RAM from Tensorflow

**Comment by ch-hristov** [comment]
I'm feeling pain from this issue 2 years later again. For the love of god we need a fix for this

**Comment by vineethdas** [comment]
I have a similar issue when running the inference model with Celery worker, the allocated GPU remains there in the memory even after program has completed successfully.

**Comment by DanielWicz** [comment]
Any progress ? It's not amusing any more

**Comment by vineethdas** [comment]
Using the numba libraries helped me aswell, normally tensorflow 2 automatically closes the sessions, unlike tensorflow1.
But while working with celery worker nodes, the sessions are not closed, as the PID holds onto celery somehow, manually including cuda.close() from numba might help. It did for me :)

**Comment by DylanHsu** [comment]
Hello, is there any plan to add functionality for clearing GPU memory to Tensorflow? At best, this is a really embarrassing design oversight.

**Comment by 3a2l** [comment]
Same problem and I have come to the same conclusions. Additionally I am using the C API (inference) which is not as capable as feature rich as the Python one. I ended up patching and building TensorFlow myself, where I have exposed two functions that are meant to be used for testing only. The functions are `GPUProcessState::TestOnlyReset` and `ProcessState::TestOnlyReset`. This resets the singleton allocator objects that control the memory allocations on the CPU and GPU (CUDA) - a `Session` needs to be recreated after this, which I don't mind as it fits my use case.

This is far from a clean solution but it works for me and honestly issues/tickets on this platform for this project are next to worthless.

**Comment by steemandlinux** [comment]
```
import gc
gc.collect()
```
Working nice with AMD.

**Comment by ajan1019** [comment]
We are using 6 pytorch models in production. We have an easy solution for this issue in pytorch. Recently we tried tf2 for a use case. Now we want to deploy it into GPU and this issue is severely affecting us.

TF models clearly take more memory than what is needed in inference. This gives a headache in maintaining our models available in production. I wish I should have stuck with pytorch. "TF is better suited for production" looks like a joke to me.
We are using pytorch in production for almost 2 years, Never faced a problem in memory management.

**Comment by stevenwong** [comment]
Not sure if this will help but this is what I've noticed.

I'm working on time-series problems where I train a model, roll forward one month, train a model... I was initially using RTX 2060 and TF 2.1 in Windows and everything worked fine. I didn't notice any memory leak/OOM issues. Built my new PC with a RTX 3070, Ubuntu 20.04, TF 2.5, CUDA 11.2, cuDNN 8.1 (I followed the TF compatibility table so there shouldn't be an issue from a version POV). I noticed TF seems to be leaking 3-5mb (memory usage was creeping up in `nvidia-smi`) at each iteration. This really adds up after hundreds of iterations. Then I tried various settings and I noticed that if I turn JIT/XLA off, the memory leak seems to go away. Note that I wasn't using XLA in Windows with the RTX 2060 when I could train thousands of iterations without issues. So if you're having memory issues in production, do you have XLA on?

This is how I'm clearing sessions after each iteration:
```
del model # tried model = None, it doesn't have a difference
tf.keras.backend.clear_session()
gc.collect()
```

**Comment by michaellee1** [comment]
Any update?  It's really painful to have done all the legwork training in Tensorflow only to see that so much as loading a model into GPU memory basically removes the ability to swap off of that GPU during a long-running process.

This affects anyone who is running a server that:
- Also does GPU rendering/processing.
- Runs multiple models (ie Pytorch models) and would like to swap them.
- Anything else using the GPU.

I'd argue that these are critical use cases in production.

**Comment by Avelina9X** [comment]
Still having this issue. Just trying to move some 'leaky' code inside a function to hope this hints to the gc that artefacts created during the function should be cleaned up... AND IT WORKED!

can confirm this works
```py
def leaky_function():
    . . . leaky stuff . . .
    return something
thing = something()

tf.keras.backend.clear_session()
gc.collect()
```
But... this shouldn't be required. Clearly there is some issue in tensorflow making gc.collect() mandatory as well.

**Comment by brendano257** [comment]
@Avelina9X 

I am not seeing similar results with an RTX 3060 on Manjaro (Nvidia 465.31 and CUDA 11.3). I've wrapped all code in a main function, and have not been able to get TF to release any vram after the function runs, even with:

`
if __name__ == '__main__':
    main()
    tf.keras.backend.clear_session()
    del tf
    gc.collect()
    
    # try to do something else with GPU
`

Subsequent uses (in the same Python process) fail to allocate GPU memory, and I can confirm that TF is still holding onto some 10GB of vram via nvidia-smi.

If your version is in fact working for you, the manual call to `gc.collect()` isn't particularly surprising. Python's GC runs only periodically, and does not automatically remove items, even if they're marked for deletion. For instance `a = 10; del a;` only guarantees that no subsequent lines of code have access to the variable `a`, but makes absolutely no promises about if/when the memory allocated for `a` will be released. For a short-running process or one that uses little memory, it may never be released until the process terminates.

All that said, I think the simplest fix available is to run *everything* as a subprocess, such that no large amounts of data like a training set are passed to the subprocess, which is what causes the performance degradation others have mentioned.

`
if __name__ == '__main__':

    p = multiprocessing.Process(target=main)
    p.start()
    p.join()

    # use the GPU for something else now (works for me on Manjaro w/ an RTX 3060)
`
The above is taken from: https://stackoverflow.com/questions/55479221/how-to-clearing-tensorflow-keras-gpu-memory

**Comment by ZhongxuanWang** [comment]
I walked through the whole page, tried mostly every solution, and it still doesn't work.

**Comment by mushaffarasyidridha** [comment]
try tf.keras.backend.clear_session() as it will del the model that has been occupying memory.
not explicitly flush memory till empty but at least reducing model in memory.
"If you are creating many models in a loop, this global state will consume an increasing amount of memory over time, and you may want to clear it. Calling clear_session() releases the global state: this helps avoid clutter from old models and layers, especially when memory is limited."

**Comment by Pedrohgv** [comment]
I'm facing this same issue. I load a model and want to continually do predictions with it, so clearing the entire memory is out of the question (it would also clear the model, which I want to load only once at the start of the application and then have it running predictions indefinitely).
What amuses me is the fact that the memory usage grows after some batches pass through the model. From my understanding, if I have only one model loaded, the memory should stay constant no matter how many predictions I run on it, but it seems TensorFlow somehow just keeps increasing the memory usage as predictions happen.

**Comment by stevenwong** [comment]
Have you guys tried my suggestion of turning XLA off? `tf.keras.backend.clear_session()` works for me after turning XLA off and I can repeatedly train models (in fact, I've been training hundreds of new models) without restarting python to clear memory.

**Comment by Pedrohgv** [comment]
So I found out that my issue was that I was using [decord](https://github.com/dmlc/decord) for loading videos and setting `decord.bridge.set_bridge('tensorflow')` (which allowed frames to be loaded in tensorflow tensors). Removing this and then transforming the values with `tf.constant()` fixed the leak.

**Comment by sanjoy** [comment]
Hi everyone,

It looks like there are couple of different issues here, with different root causes & fixes.

1. In some cases the TensorFlow process does not exit ("the only way I can then release the GPU memory is to restart my computer").  I suspect this behavior is environmental and is outside TensorFlow so we don't have a fix or a workaround.  If you're able to share a reproducer for this that we can run (e.g. on a cloud machine or locally) then I'm happy to take a look.
2. TensorFlow allocates almost all the memory on the GPU by design, which does not play well with other processes trying to use the same chip (https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-871755065).  This is a deliberate choice, but we have a workaround for now, and are working towards a longer term solution.  More on this below.
3. Something specific to how XLA GPU integrates with TensorFlow (https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-903189335).  **I'd suggest filing a separate issue for this with a reproducer**, since I believe this is distinct from (1) and (2).


Regarding (2), there are a couple of options:

1. Enable `allow_growth` (e.g. by adding `TF_FORCE_GPU_ALLOW_GROWTH=true` to the environment).  This will prevent TF from allocating all of the GPU memory on first use, and instead "grow" its memory footprint over time.  However, there are a few caveats: a) this does not cause TF to release memory (i.e. we only grow, we don't shrink) so it may not help in some use cases, and b) it increases fragmentation so models that used to fit can now run OOM.
2. Enable the new CUDA malloc async allocator by adding `TF_GPU_ALLOCATOR=cuda_malloc_async` to the environment.  This leverages the new async allocation APIs, which allows us to have BFC-like behavior without preallocating all of the memory on the GPU.  This is experimental currently (and the implementation was contributed by @nouiz from NVIDIA), but we're working towards making it non-experimental.

**Comment by robotoil** [comment]
How about an example of how to properly release memory?

**Comment by nouiz** [comment]
About cudaMallocAsync. The verison in TF isn't working.
This PR fix it: https://github.com/tensorflow/tensorflow/pull/50961 So watch this PR and when it get fixed, wait a day and you can try TF nighly.

@robotoil it is currently not possible to ask TF to release memory.

@sanjoy About allow_growth, you are right that in theory it can still increase memory fragmentation. But recently it was changed to use the cuMemMap interface and grow an allocation inside the same virtual address. It doesn't create distinct poll as before, only 1 pool that grow.
So the fragmentation now is much lower and to my knowledge only come from the ptr being at different position, so it trigger different fragmentation issue. Do you know a real case where the current allow_growth implementation trigger a new OOM, while not using allow_growth works?

**Comment by nouiz** [comment]
The fix fpr cudaMallocAsync was merged a few hours ago.
Can you wait 24h and try TF nightly build to be sure that it also works for you?
If you have any comments on that new features, please share with us.

To enable it, use the environment variable `TF_GPU_ALLOCATOR=cuda_malloc_async`.

Please share your result on this new feature.

Note, with this new allocator, the cuda driver will release automatically reserved, but not used memory when other library in the same process make a cudaMalloc calls that miss memory. So you do not need to trigger a trim command yourself.

**Comment by jiangyangzhou** [comment]
This problem is radicurous, I saw this problem in TF1.x era (2016), and when it comes TF2 in 2021, this problem still makes no progress. People struggle to solve this problem, but all the solutions look ugly.  I am sure I would still be a Pytorcher, and I would never use Tensorflow. (Meet trouble just for copying a single model...)

**Comment by nouiz** [comment]
The current proposed solution will hopefully before the default in the future. It is a new memory allocator.
What do you find ugly with that solution? I'm very interested to know why you find it ugly as I do not see it like this.

The environment variable is something temporary to help us test the new memory allocator before doing changing the default. Changing a memory allocator must be done with caution as it have many implication. This is why it is available as an option first. If this is what you find ugly, it should go away at some point.

**Comment by Eli177** [comment]
Setup:

TF 2.5
Cuda 11.4
Nvidia RTX 2080 Super.

Trying to run a bunch of tests that involve building models, running a training and evaluation step. Each model is built in series in its own subprocess which terminates, after calling 
`del models
tf.keras.backend.clear_session()
 gc.collect()
 `
 The memory of the GPU is not being released after the subprocess exits, so the next model cannot be built with a slew of tensorflow/core/common_runtime/bfc_allocator.cc:1054] errors appearing when the new subprocess is started.
 
After setting TF_GPU_ALLOCATOR=cuda_malloc_async I get the following error:

```
in setup_gpus
    logical_gpus = tf.config.experimental.list_logical_devices("GPU")
venv/lib/python3.6/site-packages/tensorflow/python/framework/config.py:452: in list_logical_devices
    return context.context().list_logical_devices(device_type=device_type)
venv/lib/python3.6/site-packages/tensorflow/python/eager/context.py:1395: in list_logical_devices
    self.ensure_initialized()
venv/lib/python3.6/site-packages/tensorflow/python/eager/context.py:525: in ensure_initialized
    context_handle = pywrap_tfe.TFE_NewContext(opts)
E   tensorflow.python.framework.errors_impl.InternalError: No allocator statistics
```

Please just fix all of the memory leaks and memory allocation problems. I just want to be able to use your API without it constantly falling down.

**Comment by nouiz** [comment]
Hi, I understand that memory problems aren't great. Just a few notes:
- cudaMallocAsync isn't made to help the problem you encounter. It helps only for some other cases.
- You need TF nightly build to use cudaMallocAsync for now. The working version will be available in the next TF release.

About your issues. I tried what you describe years ago with another software then TF. It was something hard to do due to python. The only reliable way that I found is to have bash script that launch all the jobs or to launch sub process one after the other. You said that you used subprocess, but are you sure it wasn't multiple threads? They aren't the same thing and in your case could make a big difference. Also make sure that you read the subprocess return value.

**Comment by Eli177** [comment]
The processes are sequential, the next model is not built until after the previous one has finished the training and evaluation step. I call subprocess.run(my_training_script.py) which is a blocking call, i.e. the next call cannot occur until the subprocess has finished. 
Tensorflow is just not deallocating memory, even after processes finish. In order to clear the gpu memory I have to kill the process that is spawning subprocesses.

**Comment by chenghuige** [comment]
from numba import cuda
   cuda.select_device(0)
   cuda.close()
This works on kaggle kernel, when I first run tf2 keras predict then release gpu and do torch inference.

**Comment by andreped** [comment]
I'm using TensorFlow in C++ and I am observing the same. 

In Python, I have successfully created background processes for performing inference and that works well, as the session is created and deleted in the background process, keeping the main process free of this memory leakage issue.

However, in C++ this solution does not seem to work, as creating processes in C++ is not viable. It really needs to be a proper reset method of sorts for this.

Would love if TensorFlow could fix this really annoying issue... It is making it challenging to use TensorFlow in production!

**Comment by DanielRudnicki** [comment]
> from numba import cuda cuda.select_device(0) cuda.close() This works on kaggle kernel, when I first run tf2 keras predict then release gpu and do torch inference.

This worked for me:)
TF version 2.7.0, docker image

**Comment by LoveDH** [comment]
> from numba import cuda cuda.select_device(0) cuda.close() This works on kaggle kernel, when I first run tf2 keras predict then release gpu and do torch inference.

This worked for me !
TF version 1.15.0--gpu in Colab

**Comment by jvishnuvardhan** [comment]
Looks like the issue got resolved by the resolution provided by @chenghuige . Can we close this issue? Thanks!

**Comment by 3a2l** [comment]
No, this does not resolve the issue. What we are asking for is a clear and clean TensorFlow API function (some of us need it in the C API as well) to clear the memory TF allocated.

**Comment by andreped** [comment]
I agree with @3a2l. This has **not** been resolved. Need a proper, stable solution for freeing GPU memory through the TF API. This is relevant for all programming languages that use TF. Strange that this does not exist in 2022.

**Comment by FirefoxMetzger** [comment]
I'm going to add my 2 cents here, hoping that this will be useful to some, but also knowing that this goes against the currently popular opinion.

Just because `nvidia-smi` says that TF occupies (almost) all available GPU memory, doesn't actually mean that all of this memory is actively in use. TensorFlow manages device memory by itself and what `nvidia-smi` reports is the amount of memory that is currently under TensorFlow's management rather than the amount of memory that is occupied by tensors. 

If you want to see how much memory is currently occupied by tensors use: [tf.config.experimental.get_memory_info](https://www.tensorflow.org/api_docs/python/tf/config/experimental/get_memory_info). This is because

> For GPUs, TensorFlow will allocate all the memory by default, unless changed with [tf.config.experimental.set_memory_growth](https://www.tensorflow.org/api_docs/python/tf/config/experimental/set_memory_growth). The dict specifies only the current and peak memory that TensorFlow is actually using, not the memory that TensorFlow has allocated on the GPU.

If you are running into OOM exceptions, chances are high that you (unknowingly) kept a reference to your model around which prevents that memory from being freed. Here is a list of a few pitfalls that are easy to fall into (and that I _definitely_ didn't learn about the hard way):

- In Jupyter (IPython), if your training loop (or call to `model.fit`) uses objects from different cells, make sure those objects don't keep references to your model or any tensors. If they do, those tensors will stick around.
- The `history` object returned by `keras.model.fit` holds a reference to both the model and its callbacks. If you keep the history object around the model will stick around, too.
- Callbacks (`tf.keras.callbacks.Callback`) hold a reference to the model (`self.model`), so your model will live as long as the callback lives. The same applies if you are using custom callbacks that hold references to sub-models (or other things that reference the Graph).
- `tf.data.Dataset` may be (partly) allocated in the GPU and will become part of the graph during training. As such, it will be alive for as long as your model is alive.

---

Most of the time your main process will be the only thing that uses the GPU, so it is okay for TF to use all available vRAM. Sometimes, however you may want tensorflow to manage less memory (because you have multiple processes that need parallel access to the GPU). IN this case use a [logical GPU](https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth) and limit its memory to the peak amount that the current process needs.

---

**TL:DR**: Intuition on why allocating all the memory is actually desirable.

TF allocating (and hogging) _all_ available GPU memory from the start is actually a good thing and exactly what you would expect from a managed solution like this. Conceptually, this is one of the big differences between TF and torch. Torch asks you to manage your own memory (`.to(device)` and consorts), which gives you greater power, but - as it goes - with greater power comes greater foot shooting potential. You have to worry about things like memory fragmentation (in long-running processes), the performance impact of placing batches (or the entire dataset, if you are okay with small models) onto the GPU or the host, how many batches to place into the GPU, how to distribute if you are using multiple GPUs, etc. TF will worry about all of this for you, but that comes at the cost of giving TF an aligned chunk of memory that it can work with. In a nutshell, that's what you are buying when you go down the managed route and why it is not a bad thing that TF holds on to the memory until you terminate the process (giving memory back incrementally is a potential source of aforementioned memory fragmentation).

**Comment by andreped** [comment]
@FirefoxMetzger I feel that you misunderstand what is actually the issue here. If you were to load models (and clear the graph and delete the relevant objects) multiple times sequentially, you will run into OOM using TF. That demonstrates the core problem. I'm not talking about allocating some or all memory. OOM will occur eventually regardless as some memory will remain for each iteration.

I believe this happens because TF creates the session globally and fails to have a proper clear method, which removes all necessary objects. Might be that we are just unable to remove them through the API. We have the exact same problem in C++ using TF, and I have found no way of fixing that. I believe it is a fundamental issue with the way sessions work in TF.

In production this is a **big** problem. The only solution I have used is to use multiprocessing and wrap the entire TF process inside independent processes, which is not ideal and not possible for all applications.

I can understand why there might be some "leak" due to any number of reasons, but the leak I am observing is quite severe. I also don't buy the "that is what you get when you use TF" argument. If you work with torch you will come to realize that it is actually not that low-level and challenging to use. Torch has come a long way. The main reason why I put my energy on TF was with the introduction of Keras, but if you compare TF today with PyTorch-Lightning (for instance), there really is not that much difference. However, we have several projects which depends on TF and numerous developers that use it in production. 

This is indeed a **big** issue and we cannot hide it under the rug!

**Comment by DylanHsu** [comment]
@FirefoxMetzger This is probably the worst instance in the steady parade of “TF apologists” over the past two years, of various individuals who do not understand the issue claiming it is not a problem. As explained above, it is a real issue in production. At this point, it has become amusing to receive a new email from another apologist every few months, but what is not amusing is that this repeatedly derails the conversation and the problem never gets solved.

**Comment by nalane** [comment]
Seriously, try doing hyperparameter optimization using nested cross validation. You WILL run out of GPU memory in TF eventually. No amount of `delete model` or `tf.config.experimental.set_memory_growth(gpus[0], True)` or `tf.keras.backend.clear_session()` or `gc.collect()` has ever worked for me. Even in cases when I don't keep the `history` object, when I don't use callbacks, and when I don't use `Dataset`s (I typically just use Python lists), this is still an issue for me.

This has been a problem for two years now. Every once in a while, I'll come back to this thread to see if anything's been done, but nope. This is never getting fixed. Just use PyTorch.

**Comment by FirefoxMetzger** [comment]
@andreped You're right, I think I have indeed misunderstood the main issue. Let's see if I get it right the second time around: It is not so much that tensorflow allocates a big block of memory from the start, but rather that tensorflow's leaks GPU memory?

I actually never ran into that particular issue, but it does indeed seem to be the case. Here is a snippet that - I think - visualizes the problem:

```python
import tensorflow as tf
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker

@ticker.FuncFormatter
def sizeof_fmt(num, pos, *, suffix="B"):
    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
        if abs(num) < 1024.0:
            return f"{num:3.1f}{unit}{suffix}"
        num /= 1024.0
    return f"{num:.1f}Yi{suffix}"


def log_usage(skip=False):
    foo = tf.keras.Input((1000,), dtype=tf.uint8)
    if not skip:
        x = tf.keras.layers.Dense(1)(foo)
    return tf.config.experimental.get_memory_info('GPU:0')["current"]

fig, ax = plt.subplots()
ax.plot(
    [tf.config.experimental.get_memory_info('GPU:0')["current"]]
    +[log_usage() for _ in range(250)]
    +[log_usage(skip=True) for _ in range(250)]
)
ax.yaxis.set_major_formatter(sizeof_fmt)
ax.set_xlabel("Num Re-Instantiations")
ax.set_ylabel("Used vRAM")
fig.tight_layout()
fig.savefig("memtest.png", dpi=150)
```

![memtest](https://user-images.githubusercontent.com/4402489/159140148-f6348a32-4c24-4cf3-86a5-583273ad3ccd.png)

What I would have expected to see is that after the GC ran the usage is reset back to zero (or at least to some small value that is independent of the number of re-instantiations). Also - to my surprise - it looks like tensorflow doesn't run GC once the current scope exits (CPython's policy), but instead runs at what appears to be regular intervals. Further, if I change the layer size, memory usage (and the amount of leakage) increases:
![memtest](https://user-images.githubusercontent.com/4402489/159140368-a01feed7-4c19-4f23-bc4a-99619b7662ca.png)
(I increased the number of input units 100x and I end up leaking 100x more memory. If I keep increasing the number of parameters (or increase the number of re-instantiations), I will eventually see an OOM exception.)

I ran this on TF 2.8.0 on both Windows and Linux, on 3 different GPUs (with different CUDA versions), and with both the BFC memory manager and the experimental async allocator (`TF_GPU_ALLOCATOR=cuda_malloc_async`); results are consistent across all variations I've tried. Assuming this does capture the core issue here, perhaps it can be a starting point to track down where it comes from.

**Comment by nabilapuspit** [comment]
is there someone who can help me? I am very new to TF and AI and got stuck in this problems for a several weeks.

Ubuntu 18.04
installed from source (with pip)
tensorflow version 2.6.2
CUDA 11.6
cudnn-11.3-linux-x64-v8.2.1.32
GeForce GTX 1080 Ti, 32GB RAM

I've tried to run my training model and the model can run perfectly for 14 epoch with batch size 100. but then the next validation step needs models to save as .pd files, so I tried to save the model as .pb files and after that my training models can not running again and got stuck at epochs 12. although I tried to restart the computer and delete the saving .pb files steps it still didn't help and still got OOM and [CUDA_ERROR_OUT_OF_MEMORY] problem. pls anyeone:')

**Comment by ctaoist** [comment]
> This has been a problem for two years now. Every once in a while, I'll come back to this thread to see if anything's been done, but nope. This is never getting fixed. Just use PyTorch.

I am running into this problem in jupyter notebook, but I have not used pytorch. Excuse me, does pytorch has no this problem?

**Comment by j-mendez** [comment]
When using node.js it's almost impossible to clear the memory being held. It's not realistic to start a new sub process in node either to handle each request. If you try to clear any model including the default Mobilenet and CocoSSD etc, no luck. This issue is not related to gpu entirely and occurs on `wasm` backends too.

**Comment by nalane** [comment]
I’m not sure how node.js or wasm has anything to do with any of this

**Comment by donglinz** [comment]
Any update on 2022?

**Comment by andreped** [comment]
> Any update in 2022?

Doubt it will ever be fixed. Perhaps in TF 3, but doubt it 📦 

I guess for now, don't use TF in deployment. Use OpenVINO/TensorRT with the ONNX format instead, but for training models TF works just fine.

**Comment by alexcoca** [comment]
:(

**Comment by qmarcou** [comment]
I'm also facing this OOM issue during hyperparameters tuning using keras-tuner although by largest possible model can fit on the GPU alone (with a large margin in fact). 
@FirefoxMetzger thanks for trying to build this minimal example and taking this forward! I've built on it to try and nail more precisely the issue.

Unfortunately @FirefoxMetzger minimal exemple does not reproduce the issue discussed here, by forcing garbage collection and session clearing, and logging the max GPU memory usage (instead of current) which I would think is more appropriate to look for an OOM event:
``` python
import tensorflow as tf
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import gc
from tensorflow import python as tfpython

@ticker.FuncFormatter
def sizeof_fmt(num, pos, *, suffix="B"):
   for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
       if abs(num) < 1024.0:
           return f"{num:3.1f}{unit}{suffix}"
       num /= 1024.0
   return f"{num:.1f}Yi{suffix}"


def log_usage(skip=False):
   # session cleanup
   tf.keras.backend.clear_session()   
   gc.collect()
   # Reset memory stats 
   tf.config.experimental.reset_memory_stats('GPU:0')
   foo = tf.keras.Input((1000,), dtype=tf.uint8)
   if not skip:
       x = tf.keras.layers.Dense(100)(foo)
   return tf.config.experimental.get_memory_info('GPU:0')["peak"]

fig, ax = plt.subplots()
ax.plot(
   [tf.config.experimental.get_memory_info('GPU:0')["peak"]]
   +[log_usage() for _ in range(250)]
   +[log_usage(skip=True) for _ in range(250)]
)
ax.yaxis.set_major_formatter(sizeof_fmt)
ax.set_xlabel("Num Re-Instantiations")
ax.set_ylabel("Max Used vRAM")
fig.tight_layout()
fig.savefig("memtest.png", dpi=150)
print(tfpython.client.device_lib.list_local_devices())
```

<img src="https://user-images.githubusercontent.com/18257721/177146364-15d2218d-bc0f-47ad-9187-840580b114a1.png" height="400" />

No problem here!

Now following what has been said in this thread and my experience with the issue, I though maybe the leak only happens upon fitting a model:
``` python
import tensorflow as tf
import tensorflow.keras as keras
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import gc
from tensorflow import python as tfpython
import numpy as np

n = 32
input_size = 1000
output_size = 100

# create some random uninformative data
x = np.random.uniform(low=-.5, high=.5, size=(n, input_size))
y = np.random.binomial(1, p=.5, size=(n, output_size))

@ticker.FuncFormatter
def sizeof_fmt(num, pos, *, suffix="B"):
    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
        if abs(num) < 1024.0:
            return f"{num:3.1f}{unit}{suffix}"
        num /= 1024.0
    return f"{num:.1f}Yi{suffix}"


def log_usage(skip=False):
    tf.keras.backend.clear_session()
    gc.collect()
    tf.config.experimental.reset_memory_stats('GPU:0')
    if not skip:
        model = keras.Sequential([
             keras.layers.InputLayer(input_shape=(input_size,),
                                     dtype=tf.float32),
             keras.layers.Dense(output_size, activation="sigmoid"),]
        )
        model.compile(loss='binary_crossentropy')
        model.fit(x=x,
                  y=y,
                  batch_size=32,
                  epochs=1,
                  verbose=False)
    return tf.config.experimental.get_memory_info('GPU:0')["peak"]

fig, ax = plt.subplots()
ax.plot(
    [tf.config.experimental.get_memory_info('GPU:0')["peak"]]
    +[log_usage() for _ in range(100)]
    +[log_usage(skip=True) for _ in range(100)]
)
ax.yaxis.set_major_formatter(sizeof_fmt)
ax.set_xlabel("Num Re-Instantiations")
ax.set_ylabel("Max Used vRAM")
fig.tight_layout()
fig.savefig("memtest_fit.png", dpi=150)

print(tfpython.client.device_lib.list_local_devices())
```

<img src="https://user-images.githubusercontent.com/18257721/177146746-235947c3-4cf2-4bc8-90fa-f86b81798b93.png" height="400" />

Again no problem here, and peak memory usage is ~2.5 x higher during training due activations and gradients

Both test have been run on the same K80 GPU with compute capability 3.7, and it doesn't not support the `cuda_malloc_async` (requires compute capability >=6.0) hence it was not used.

Anybody has an idea on a faulty ingredient to add to this MWE to reproduce the issue discussed above?

**Comment by qmarcou** [comment]
@FirefoxMetzger the point of `tf.config.experimental.reset_memory_stats('GPU:0')` was to be able to measure peak memory usage on a given time interval. I think using `tf.config.experimental.get_memory_info('GPU:0')["current"]` does not represent what we want to measure here, we really want to see the maximum memory usage, since this is what will cause the OOM error.

The thing is, if you remove the call to `tf.config.experimental.reset_memory_stats('GPU:0')`, you will indeed get the peak memory consumption since the beginning of the program, but this will, by definition, be an increasing function over time. You then have no way to assert whether this increase over time is an allocation fluctuation (though programming should be somewhat deterministic) or an actual memory leak. 
The point of reseting the memory counter right after GC is to reset it, a priori, to the min memory usage.
If there were to be a memory leak, we would expect both the value at resetting (not logged at the moment) and the 'peak' value to increase over time.

**Comment by FirefoxMetzger** [comment]
> we really want to see the maximum memory usage, since this is what will cause the OOM error.

Yep, I agree with your general logic here, and for long-running programs, this is indeed the better metric to track. 

For the program that I shared here specifically, `current` and `peak` are the same and there shouldn't be a difference if you replace `tf.config.experimental.get_memory_info('GPU:0')["current"]` with `tf.config.experimental.get_memory_info('GPU:0')["peak"]`. Both variants should show the leak.

---

> but this will, by definition, be an increasing function over time. You then have no way to assert whether this increase over time is an allocation fluctuation (though programming should be somewhat deterministic) or an actual memory leak.

Could you explain why this should be the case?

In my mind, if I write a program that continuously allocates and frees a block of memory then a good memory manager should recognize this and reuse the previously freed block instead of continuously requesting new blocks and fragmenting the memory in the process. This might not happen immediately depending on how one chooses to do garbage collection (hence the sawtooth pattern in the figures I shared), but it shouldn't grow without bounds. Further, at the end of the program once everything has been freed and it essentially just runs idle (when we call `log_usage(skip=True)`) memory usage should fall back to the initial value at the start of the program because all the memory should be returned sooner or later because it is no longer in use. This is the step that we see around iteration 400; however, instead of dropping to 0 it drops to the valley of the most recent sawtooth. Anything below that line is memory that was leaked. We don't hold any reference to it and, hence, can't use it, but GC refuses to free it so we can't re-request it either.

Ideally, I would like to see a constant line, because - IIRC - TF promises to honor python's GC so there shouldn't be any accumulation of allocated memory. Personally, I'd also be happy to see a clean sawtooth that falls back to zero every couple of iterations (GC is expensive and if we really need it every iteration we can manually request it). The problem is that there is a sawtooth + a trend line, which - to me - is unexpected and I can't explain it other than through a memory leak.

**Comment by qmarcou** [comment]
> > but this will, by definition, be an increasing function over time. You then have no way to assert whether this increase over time is an allocation fluctuation (though programming should be somewhat deterministic) or an actual memory leak.
> 
> Could you explain why this should be the case?

From a mathematical perspective you can define peak_mem(t) as peak_mem(t)=max(current_mem(t),peak_mem(t-1)), thus peak memory is bound to be an increasing function (in the mathematical sense, meaning constant or increase as opposed to strictly increasing). Sorry if that wasn't clear.
I put more trust on this mathematical property than on my understanding of how GC works and how it is actually implemented, hence my comment.

> In my mind, if I write a program that continuously allocates and frees a block of memory then a good memory manager should recognize this and reuse the previously freed block instead of continuously requesting new blocks and fragmenting the memory in the process. This might not happen immediately depending on how one chooses to do garbage collection (hence the sawtooth pattern in the figures I shared), but it shouldn't grow without bounds.

In theory I fully agree with you, but again I would refrain from making hypothesis on how GC and malloc work and there might be situations where memory allocation is not optimal in terms of memory usage (but maybe it is in runtime, or simply it isn't in general because it's crazy hard to design a perfect malloc system.
This is also why I actually force garbage collection, to eliminate some scheduling details of python's GC that are not relevant here (I think).  

> Further, at the end of the program once everything has been freed and it essentially just runs idle (when we call `log_usage(skip=True)`) memory usage should fall back to the initial value at the start of the program because all the memory should be returned sooner or later because it is no longer in use. This is the step that we see around iteration 400; however, instead of dropping to 0 it drops to the valley of the most recent sawtooth. Anything below that line is memory that was leaked. We don't hold any reference to it and, hence, can't use it, but GC refuses to free it so we can't re-request it either.

But here you make strong assumptions on what Python's or TF garbage collection should look like :)
And I think that Python has a much more lazy approach to garbage collection than you seem to think. Indeed GC is expensive, and I think Python will only perform GC when needed, ie when allocating new objects that would otherwise increase memory usage unnecessarily. I think this is also linked to the fact that requesting memory to the OS is expensive languages such as python and java prefer to keep the memory they have obtained from the OS. Since in your code snippet  the `idle` portion, calling `log_usage(skip=True)`,  is not creating any new object Python has no reason to worry about the amount of memory it retains, especially when it's as small as a few MB, which is very little on current hardwares.
That being said I'm far from being expert on all these memory inner workings, and I may be completely wrong.

> Ideally, I would like to see a constant line, because - IIRC - TF promises to honor python's GC so there shouldn't be any accumulation of allocated memory. Personally, I'd also be happy to see a clean sawtooth that falls back to zero every couple of iterations (GC is expensive and if we really need it every iteration we can manually request it). The problem is that there is a sawtooth + a trend line, which - to me - is unexpected and I can't explain it other than through a memory leak.

But isn't that exactly what my graphs are showing? Once your force garbage collection the memory requirement simply drops to 0, and peak usage is constant before that.
But maybe we don't agree on the 'memory leak' definition? In my understanding a leak is memory that the program requests and lose track of (e.g in C++ you would call `new int` but forget to register the pointer to the allocated memory address of that int and hence cannot delete it afterwards). Here if the GC is able to free up the memory, then it means it has not lost track of instantiated objects, hence no memory leak.
For me the two graphs I have posted show exactly what I expect from a non leaky program. That being said, I myself had (and still have) issues with some kind of memory leak for which I have to use the subprocess/multiprocessing trick, and I just think that our minimal example is not sufficient to reproduce the issue and that the leaky ingredient is still missing.

**Comment by qmarcou** [comment]
@FirefoxMetzger great! I was also wondering whether `tf.keras.backend.clear_session` had any effect, and I guess we can safely conclude that it's useless in this context. 
I've realized that our plots may be masking small leaks due to scaling, and I've rerun the previous experiment plotting on log scale, and running a "pre work" IDLE time before actual work (Dense layer instantiation or model fitting) to get a clean baseline.
Left plot shows pre IDLE-work-post IDLE memory usage, the right plot show only memory requirement  

For layer instantiation here's what I get:
![memtest_instantiate_100unit](https://user-images.githubusercontent.com/18257721/177750631-3967a3aa-e6ca-4f6e-b045-61568b18464a.png)



For the model fitting version:
![memtest_fit_100_32_100_1000](https://user-images.githubusercontent.com/18257721/177751009-707926bb-9d47-4173-b519-e2d04fba86a4.png)

*I'll try and add the code later I don't have the time to do it right now*

From these plots I think there are 2 takeaways: there's a small leak (hidden in our previous plots), and it's more important when you compile and fit the model (768B for the layer instantiation and 2.5KiB for the fit). 
One thing I do not understand is also the small increase in memory requirement between the first and second run for both versions ( 256B for the layer instantiation and 152.75KiB for the fit version). Maybe it's a JIT thing (random suggestion)? 
I've tried to vary the configurations a bit by hand and did not find a parameter that would change the 2 numbers above for the layer instantiation version.
For the fit version the batch size and number of units seem to change the amount by which memory usage increases between the first and second run, but did not seem to influence the 2.5KiB final leak. 
I guess we could make this analysis more systematic trying to vary: the number of batches seen in the fit (if this increases the amount of leak this would explain why calling predict repeatedly as mentioned above could lead to OOM), the batch_size, input size, output size, number of runs.


Some remarks on your last code example: I guess you added the child Process usage to make the different runs independent  in terms of GPU memory (which is great!) but I don't think that your design is sufficient to do so (or at least may not be very robust to small changes in code or OS). See #8220 and #5448. The simplest change would be to move tf import into the `run_experiment` function since you do not need it outside of this function.

Maybe we could have a bit of help from the TF team now that we are trying debug this issue? @sanjoy ?

**Comment by tdat97** [comment]
if you use "Process" of multiprocessing.. Maybe you will face error such as "CUDA_ERROR_NOT_INITIALIZED".
(tf 2.x)
I did.

Generally, we used to code this such as 
"model.fit(train_x, train_y, epochs=30, validation_data=...)" 
OR
"
with tf.GradientTape() as tape:
    yhat = model(...)
    loss = loss_fn(...)
    grad = tape.gradient(...)
    opt.apply_gradients(...)
 "

I think This Way make Trainset to stack at GPUmem...
(if Trainset change again and again) <- for Data Augmentation
Therefore, it occured OOM.


---


**Then**
As Another Way, try to "tf.data.Dataset.from_generator"

```python
train_ds = tf.data.Dataset.from_generator(...)
valid_ds = tf.data.Dataset.from_generator(...)
train_ds = train_ds.shuffle(4096).batch(batch_size, drop_remainder=True).prefetch(2)
valid_ds = valid_ds.shuffle(4096).batch(batch_size, drop_remainder=True).prefetch(2)
model.fit(train_ds, epochs=n_epochs, validation_data=valid_ds)
```

---

I coded this
```python
gpu = tf.config.experimental.list_physical_devices('GPU')
tf.config.experimental.set_memory_growth(gpu[0], True)
```
for observing GPUmem.

I think this way seems no stacking GPUmem.

![image](https://user-images.githubusercontent.com/48349693/184056285-9a3ae755-cefd-4f52-a79f-31ace9dc4efb.png)


--- 

I wrote it down during training, so I can't say for sure...
Only, growth GPUmem is stoped at 9101MiB while training model.

**Comment by Gargisingh11** [comment]
Hi , What worked for me was insetad of I trying to release the GPU memory I tried to allocate the GPU memory for my model. The below link definitely helped me 
https://medium.com/ibm-data-ai/memory-hygiene-with-tensorflow-during-model-training-and-deployment-for-inference-45cf49a15688 
if gpus:
 for gpu in gpus:
    tf.config.experimental.set_virtual_device_configuration(gpu,
         [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=4096)])

**Comment by sushreebarsa** [comment]
@ HristoBuyukliev Could you refer the comment above and let us know if it helps?
Thank you!

**Comment by hmaarrfk** [comment]
Not stale

**Comment by DylanHsu** [comment]
Looks like Tensorflow is better at cleaning up GitHub issues than the memory.

**Comment by robotoil** [comment]
This is serious. Why can’t it be addressed already. Who is going to use this in production?On Oct 26, 2022, at 11:29 AM, Dylan G. Hsu, Ph.D ***@***.***> wrote:﻿
Looks like Tensorflow is better at cleaning up GitHub issues than the memory.

**Comment by nalane** [comment]
@Gargisingh11 I have not had the opportunity to try your example, but it looks conceptually almost identical to the suggestions in [this comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582318197) and [this comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582651686), as all three suggestions are focused on limiting memory growth. As @HristoBuyukliev mentions in [this response](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582324306), while that’s a nice feature, we’re talking about memory leaks originating from models that are no longer needed (e.g., in a program that is trying out multiple models in sequence).

@sushreebarsa, I get the feeling that the reason @HristoBuyukliev is not replying to your ping is that they created this issue 2 years and 8 months ago. They haven’t replied to this thread since February 2020. To put that in perspective, that was before the pandemic hit America. I’m willing to bet that they’ve moved on from TF and/or whatever they were trying to use it for in the first place. But as I’m sure you can also see, many, many people have experienced this issue in the meantime. I beg you, please look at @FirefoxMetzger’s code in [this comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-1073123668). He has some great code with graphs that explicitly show exactly what we in this thread are all begging the TensorFlow team to fix.

**Comment by hoondy** [comment]
> @Gargisingh11 I have not had the opportunity to try your example, but it looks conceptually almost identical to the suggestions in [this comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582318197) and [this comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582651686), as all three suggestions are focused on limiting memory growth. As @HristoBuyukliev mentions in [this response](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582324306), while that’s a nice feature, we’re talking about memory leaks originating from models that are no longer needed (e.g., in a program that is trying out multiple models in sequence).
> 
> 
> 
> @sushreebarsa, I get the feeling that the reason @HristoBuyukliev is not replying to your ping is that they created this issue 2 years and 8 months ago. They haven’t replied to this thread since February 2020. To put that in perspective, that was before the pandemic hit America. I’m willing to bet that they’ve moved on from TF and/or whatever they were trying to use it for in the first place. But as I’m sure you can also see, many, many people have experienced this issue in the meantime. I beg you, please look at @FirefoxMetzger’s code in [this comment](https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-1073123668). He has some great code with graphs that explicitly show exactly what we in this thread are all begging the TensorFlow team to fix.

This is a great summary.

**Comment by phrasenmaeher** [comment]
> Hi @HristoBuyukliev , this is a very old issue that everyone is facing in TF 1.x as well as TF 2.x, it seems to be a design flaw and the TF team doesn't seem to care about fixing (I have been facing this issue for more than 2 years now).
> 
> What worked well for me was just to run my train/eval in a separate process and wait for it to finish. So when the process finishes the system kills it and releases the GPU resources automatically. You can achieve this by doing something like:
> 
> ```
> import multiprocessing
> 
> process_eval = multiprocessing.Process(target=evaluate, args=(...))
> process_eval.start()
> process_eval.join()
> ```

Here's how to get the return value of the process:
```python
from multiprocessing import Process, Queue
import random

def my_func(arg, q):
    ret_val = random.random()
    print(ret_val, type(ret_val))
    q.put(ret_val)
    return 1


if __name__ == '__main__':
    queue = Queue()  # Here
    p1 = Process(target=my_func, args=('not used', queue))
    p1.start()
    p1.join()

    res = queue.get()
    print(res, type(res))
```

**Comment by nalane** [comment]
Has there been any movement on this?

**Comment by AmosDinh** [comment]
> > Hi @HristoBuyukliev , this is a very old issue that everyone is facing in TF 1.x as well as TF 2.x, it seems to be a design flaw and the TF team doesn't seem to care about fixing (I have been facing this issue for more than 2 years now).
> > What worked well for me was just to run my train/eval in a separate process and wait for it to finish. So when the process finishes the system kills it and releases the GPU resources automatically. You can achieve this by doing something like:
> > ```
> > import multiprocessing
> > 
> > process_eval = multiprocessing.Process(target=evaluate, args=(...))
> > process_eval.start()
> > process_eval.join()
> > ```
> 
> Here's how to get the return value of the process:
> 
> ```python
> from multiprocessing import Process, Queue
> import random
> 
> def my_func(arg, q):
>     ret_val = random.random()
>     print(ret_val, type(ret_val))
>     q.put(ret_val)
>     return 1
> 
> 
> if __name__ == '__main__':
>     queue = Queue()  # Here
>     p1 = Process(target=my_func, args=('not used', queue))
>     p1.start()
>     p1.join()
> 
>     res = queue.get()
>     print(res, type(res))
> ```

Does this work with gpu?

**Comment by nalane** [comment]
> Does this work with gpu?

Yeah, it does

**Comment by exgphe** [comment]
I can confirm that in CPU mode Tensorflow does not release the memory either. I created a small experiment that creates a small model, and then executes `del model`, `tf.keras.backend.clean_session()` and `gc.collect()`, in a for loop. The memory usage grows steadily and will end up occupying all of your RAM. The only way to release the memory is to create the model in a separate process and then kill the process.

**Comment by nalane** [comment]
@reedwm @sachinprasadhs @mohantym I don't want to come across as rude, but can we get a confirmation from you guys that you at least recognize this as an issue? We have provided graphs and data showing that TensorFlow has a memory leak, and yet every time someone comes along with a "solution" that doesn't fix the problem at hand, someone from Google tries to close this thread. From my perspective, it's as if Google is trying to say one of two things:
1.  "We don't even *care* that people are telling us there is a memory leak," or
2. "We know it's a problem and are trying to sweep it under the rug."

**Comment by fingoldo** [comment]
It's just unbelievable to see this in year 2023. And you guys (core tensorflow devs) call it a production-ready machine learning framework? Learn how to clean up your crap (claimed GPU memory) first. It's disrespectful to your users. I am currently in sutuation where I need to train many models (not only ANNs, but also boostings etc, some of them also using GPU) on the same fold of data, then issue predictions with each model over a range of separate big files, proceed to the next fold etc. Spawning separate thread for every such training and every prediction task, importing tf, communicating results back to the main thread seems much more silly than just calling some helper funciton to clear allocated GPU memory. If you are able to allocate memory then you must be able to also release it easily. Find courage to tisten to you users and do it.

**Comment by Sam-Seaberry** [comment]
Not a fix but I use a docker container that I pull up each time I need to evaluate a model, once the evaluation is complete I break from the main program so the docker container is stopped. This releases all GPU resources. 

Use this container to enable GPU processing in your container (or a similar container with the correct cuda version for your gpu):
`nvidia/cuda:11.8.0-cudnn8-runtime-ubuntu20.04`

The container can be pulled up with a subprocess call like:
`subprocess.run(['docker', 'start', container_name], stderr=subprocess.PIPE)`

However, to enable GPU processing within the container add the arguments:
`--gpus all`

**Comment by nalane** [comment]
@reedwm @sachinprasadhs Do you hear how ridiculous the above is? THAT is what this memory leak is making us do.

**Comment by jurahul** [comment]
I've escalated this to the TF team.

**Comment by nalane** [comment]
@robotoil Whenever someone tells me, “I’m going to start a new project in TensorFlow,” I point them to *this* exact issue and tell them to use PyTorch instead

**Comment by robotoil** [comment]
I don’t think the devs have _any_ idea how much time is wasted by users trying to come up with workarounds. Imagine if you had to reboot your phone after each call. On Aug 25, 2023, at 11:50 AM, Nathaniel Lane ***@***.***> wrote:﻿
@robotoil Whenever someone tells me, “I’m going to start a new project in TensorFlow,” I point them to this exact issue and tell them to use PyTorch instead

**Comment by xucian** [comment]
What a funny thing to have this persist 3.5 years now. for these years-old issues (a lot of companies have them), the cause it's usually something we're not seeing (usually something detrimental to them if they fix it).
Or maybe in this case they think that just starting the python in a subprocess is an acceptable solution (for me, it was at the time).

**Comment by EKami** [comment]
My first fix was to run in a subprocess. My second fix, which was more painful, was to migrate the models to ONNX and from that point, only use PyTorch for future models

**Comment by nalane** [comment]
> the cause it's usually something we're not seeing (usually something detrimental to them if they fix it).
The fact that they haven’t even recognized it as an issue, despite the fact that we have supplied code and charts demonstrating that it’s an issue, makes me think that this is the case.

**Comment by cjmcclellan** [comment]
This issue seems to have been fixed by setting `TF_GPU_ALLOCATOR=cuda_malloc_async` as mentioned in #48545, which seems to work for me on TF 2.15 and CUDA 12.3. I'm not sure why this hasn't been mentioned here yet.

**Comment by nalane** [comment]
@cjmcclellan From https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-1073123668 in this thread:

> I ran this on TF 2.8.0 on both Windows and Linux, on 3 different GPUs (with different CUDA versions), and with both the BFC memory manager and the experimental async allocator (TF_GPU_ALLOCATOR=cuda_malloc_async); results are consistent across all variations I've tried.

**Comment by andreped** [comment]
> This issue seems to have been fixed by setting `TF_GPU_ALLOCATOR=cuda_malloc_async` as mentioned in #48545, which seems to work for me on TF 2.15 and CUDA 12.3. I'm not sure why this hasn't been mentioned here yet.

@cjmcclellan I think most of us have given up at this point :P

As @nalane mentions, if you think this fixes the issue, demonstrate it through running @FirefoxMetzger's benchmark from earlier in this thread with the proposed _fix_ (https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-1073123668)

**Comment by cjmcclellan** [comment]
@andreped and @nalane thanks sorry I missed that comment. The async allocator fixed my issue, but clearly did not entirely fix the memory leak.

**Comment by ElrondL** [comment]
Yeah TF team, please fix this. While the multiprocess trick works for single GPU, how does that work if you run distributed training? Wouldn't multiprocess be more likely to mess up the distributed training? If you run distributed training + Hyperparameter optimization you would be looking at your code eating through hundreds of GBs of RAM

**Comment by takeyama0** [comment]
In my environment, specifying a strategy mitigate the problem.
For example, in the single GPU case, the following significantly but not completely reduced memory leakage.

```
import tensorflow as tf

# set strategy
strategy = tf.distribute.OneDeviceStrategy(device="/gpu:0")

# fit and evaluate under the specified strategy
with strategy.scope():
    dataset = tf.data.Dataset.from_tensors(([1.], [1.])).repeat(100).batch(10)
    model = tf.keras.Sequential([tf.keras.layers.Dense(1,)])
    model.compile(loss='mse', optimizer='adam')
    model.fit(dataset, epochs=10)
    model.evaluate(dataset)
```

**Comment by powmod** [comment]
Just came into this thread because I have the same problem and am disheartened but what I read, on a thread that spans 4 years about a major flaw.

**Comment by fingoldo** [comment]
> Just came into this thread because I have the same problem and am disheartened by what I read, on a thread that spans 4 years about a major flaw.

Switch to pytorch, my friend, no one in google seems to be taking their tensorflow product seriously.

**Comment by jobs-git** [comment]
this is problematic indeed

**Comment by 372046933** [comment]
In my experiment. `TF_GPU_ALLOCATOR=cuda_malloc_async` have performance issues. About 11% slower than BFC allocator

**Comment by Husseinshtia1** [comment]
**Business-Grade Fix for TensorFlow GPU Memory Not Releasing**

This memory issue in TF2 isn’t just an annoyance — it’s a **cost multiplier**.

In our production environments, unreleased GPU memory led to:

* ⏱️ Slower inference due to fragmentation
* ❌ Inconsistent behavior across deployments
* 💸 Increased cloud GPU billing (idling at 90% usage)

**Our fix:** isolate training/inference inside a subprocess via a lightweight SDK:

```python
from collapse_cleaner_modules import collapse_subprocess_wrapper

def run():
    import tensorflow as tf
    # model.fit() or predict()

collapse_subprocess_wrapper(run)
```

🔒 This guarantees full GPU memory release post-run.
📈 In our case, it recovered **up to 20–25% GPU usage**, **cut downtime by 40%**, and avoided unpredictable system resets.

📊 Real-world results: [[Read the case study](https://www.linkedin.com/pulse/tool-pays-itself-30-days-what-20-gpu-recovery-looks-like-shtia-jmcrf/)](https://www.linkedin.com/pulse/tool-pays-itself-30-days-what-20-gpu-recovery-looks-like-shtia-jmcrf/)

> 🧠 **CollapseCleaner**: A runtime-safe layer for TensorFlow/PyTorch memory hygiene.
> 📦 `pip install collapsecleaner`

Why restart the whole system — when you can just restart the session?

**Comment by dgalaavy** [comment]
Can someone please explain to me why that PR fixes this issue and how to clear the memory now?

**Comment by ghost** [comment]
Great to see that Google not only fails at TensorFlow, their "Copybara Service" is equally badly written. The bot incorrectly closed this issue because the PR referenced https://github.com/openxla/xla/pull/36465, which has the same ID.

**Comment by ralexms** [comment]
> Can someone please explain to me why that PR fixes this issue and how to clear the memory now?

I resolved it by switching to PyTorch
Only regret not doing it sooner

**Comment by aydenhailey78-hub** [comment]
No issues on my end

**Comment by DylanHsu** [comment]
This is still an issue, why is it closed?
